### PR TITLE
txn: fix the lock behaviours when for update is used when low resolution tso is used

### LINKED ISSUE #54684: txn: the `tidb_low_resolution` variable is used on select for update statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t(a int key);
```

```
set tidb_low_resolution_tso = 'ON';
begin;
select * from t where a = 1 for update; // or select * from t for update
```

### 2. What did you expect to see? (Required)
The `select for update` should report an error that when `tidb_low_resolution_tso` is set, the statement could not be executed.

### 3. What did you see instead (Required)
The select for update statement executes successfully.

### 4. What is your TiDB version? (Required)

Nightly with `ca7ba14da87abd060922b32c893b121c4c3d6373` commit hash

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Issue Number: close #54684

Problem Summary:
```
 Fix the lock behaviours when for update is used when low resolution tso is used.
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
 Fix the lock behaviours when for update is used when low resolution tso is used
```

### CODE REVIEW

File: pkg/executor/adapter.go [hunk 1]
```diff
@@ -581,8 +581,13 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 	isPessimistic := sctx.GetSessionVars().TxnCtx.IsPessimistic
 
 	// Special handle for "select for update statement" in pessimistic transaction.
```
Comment (MyonKeminta) [inline pkg/executor/adapter.go:583]: This comment looks better to be moved to above the `if` at line 588

**Review by zyguan — APPROVED** [review]

File: pkg/sessionctx/variable/session.go [hunk 2]
```diff
@@ -3878,3 +3879,17 @@ const (
...
 	return s.OptObjective
 }
+
+// UseLowResolutionTSO indicates whether low resolution tso could be used for execution.
+func (s *SessionVars) UseLowResolutionTSO() bool {
+	return !s.InRestrictedSQL && s.lowResolutionTSO
+}
+
+// PessimisticLockRequired indicates whether pessimistic locks are required for the following execution.
+func (s *SessionVars) PessimisticLockRequired() bool {
```
Comment (ekexium) [inline pkg/sessionctx/variable/session.go:3889]: The function name doesn't match its body without checking its usages. Is it better to rename it to something like `PessimisticLockEligibleForSelectLock`?

**Review by ekexium — COMMENTED** [review]
The others LGTM

File: pkg/sessionctx/variable/session.go (hunk 2 above)
Comment (cfzjywxk) [reply pkg/sessionctx/variable/session.go]: Changed to `PessimisticLockEligible` to indicate whether the locks should not be ignored when `for update` clause is used.

**Review by ekexium — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**